text/2017-10-06-api-layers.md

```diff
@@ -0,0 +1,206 @@
+counted reference to a `Global`, and `Handle` is a tuple of a counted reference to a `Global` and a
+`Local`:
+
+```rust
+struct Collector(Arc<Global>);
+
+struct Handle {
+    global: Arc<Global>,
+    local: Local,
+}
+```
+
+You can create a new collector by creating a new `Global` data, and add a handle to a collector by
+sharing its `Global` data.  Using a `Handle`, you can create a `Scope` and pass it to a given
+function:
+
+```rust
+impl Collector {
+    pub fn new() -> Self;
+    pub fn handle(&self) -> Handle;
+}
+
+impl Handle {
+    pub fn pin<F, R>(&self, f: F) 
+    where F: for<'scope> FnOnce(Scope<'scope>) -> R {
+        self.local.pin(&self.global, f)
+    }
+
+    ... // other methods
+}
+```
+
+Note that `Handle::pin()`, on contrary to `Local::pin()`, does not require a reference to a
+`Global`, because `Handle` already has a reference.
+
+See [the
+implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/collector.rs) for
+more details.
+
+
+## The default garbage collector API
+
+There will be the default collector and per-thread handle to the default collector:
+
+```rust
+lazy_static! { pub GLOBAL: Global = Global::new(); }
+thread_local! { pub HANDLE: Handle = Handle::new(); }
+
+struct Handle(Local);
+
+impl Handle {
+    fn new() -> Self {
+        Self { 0: Local::new(&GLOBAL) }
+    }
+}
+
+impl Deref for Handle {
+    type Target = Local;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        self.finalize(&GLOBAL);
+    }
+}
+
+pub fn pin<F, R>(f: F) 
+where F: for<'scope> FnOnce(Scope<'scope>) -> R {
+    HANDLE.with(|handle| { handle.pin(&GLOBAL, f) })
+}
+```
+
+See [the implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/default.rs)
+for more details.
+
+
+
+# Alternatives
+
+`Scope<'scope>` vs. `&'scope Scope`: Currently the type of scope is `Scope<'scope>`. It is possible
+to use more traditional reference type `&'scope Scope`, but it may incur some runtime cost of double
+indirection.
+
+`Arc<Global>` vs. `&Global`: Currently `Collector` is implemented as `Arc<Global>`, which incurs a
+runtime overhead of reference counting.  This author believes the overhead is negligible, because
+handle creation is likely in the cold path.  Using `&Global` instead of `Arc<Global>` might
+eliminate the runtime cost, but it will significantly complicates the API.
+
+
+
+# Unresolved questions
+
+for `#[no_std]` environment, a long-term plan would be separating out what is dependent on `std`,
+namely the default collector API and the `Collector` API, and what is not, namely the internal
+API. The proposed internal API is almost `std`-free except for the fact that `Global` and `Local`
+relies on the data structures in the standard library. This author believes we can finish this job
+once [the allocator trait is stabilized](https://github.com/rust-lang/rust/issues/32838).
```

Comment:
To support `#[no_std]` environments, do we want:

1. An API that doesn't allocate memory at all?
2. An API that allocates memory, but through the `Allocator` trait?

I'm guessing the answer is (2). But if that is the case, then it's trivial to implement a `Arc` that uses the `Allocator` trait. So we could still easily have the `Collector` API in `#[no_std]`, right?

Reply:
I agree that we can easily support the `Collector` API in `#[no_std]`.  I changed the text.

That makes me think more of the organization of the Crossbeam crates.  I'd like to suggest we put the internal API (if we decide to expose it) and the `Collector` API in `crossbeam-epoch`, and implement the default garbage collector in the `crossbeam` crate.  A long-term goal is implementing `crossbeam-epoch` in the `#[no_std]` environment.  And those who just want to the default collector will import `crossbeam`, and those who want to use their own collector will import `crossbeam-epoch`.

Reply:
Interesting idea.

Another thing we should consider (or at least think about) is doing something akin to Rayon's split into `rayon` and `rayon-core`.

For example, let's say that we bump the version of `crossbeam` from `0.4` to `0.5` because we've just added a new data structure to it. Both versions depend on `crossbeam-epoch="0.1"`. Now, suppose that you have your own project that depends on crates A and B. Crate A depends on `crossbeam="0.4"` and crate B depends on `crossbeam="0.4"`. It'd be great if they both used the same global `Collector`.

This makes me think that the global `Collector` shouldn't live inside `crossbeam`, but rather inside a crate that increments versions very slowly (ideally, always at version `1.0.x`). That would be `crossbeam-epoch`, or perhaps even `crossbeam-epoch-global`. This is similar to `rayon-core`.

I'll have to think about all this a bit more. :) Perhaps we should open a separate issue to discuss.